campaigns preview panics while resolving repositories

src-cli version: 3.22.0
recommended version by our deployment: 3.20.0

`repositoriesMatchingQuery: dart_functional_test_image or dart_unit_test_image`

I hit this while trying to preview a campaign spec, ~although I'm not actually sure if this is a valid query, as I was under the impression that file-contents searches didn't work for campaigns. So maybe that's the issue~ Read through the most recent docs and see a similar query in the examples, so I guess it is a valid query. Either way, a panic is probably not the expected result.

Output:

```
✅ Parsing campaign spec
✅ Resolving namespace
✅ Preparing container images  ███████████████████████████████████████████████████████████████████████████████████████████████████████
⠹  Resolving repositories
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/sourcegraph/src-cli/internal/campaigns.(*Service).resolveRepositorySearch(0xc000209640, 0x1748da0, 0xc0000b7ec0, 0xc0003df080, 0x32, 0xc000210001, 0xc0003ab410, 0xc000529198, 0x100ea8d, 0x1651fa0)
        /home/runner/work/src-cli/src-cli/internal/campaigns/service.go:490 +0x26b
github.com/sourcegraph/src-cli/internal/campaigns.(*Service).ResolveRepositoriesOn(0xc000209640, 0x1748da0, 0xc0000b7ec0, 0xc000529280, 0x147a5e5, 0x8, 0x16b8ac8, 0xc0001ff3e0, 0x1)
        /home/runner/work/src-cli/src-cli/internal/campaigns/service.go:407 +0x192
github.com/sourcegraph/src-cli/internal/campaigns.(*Service).ResolveRepositories(0xc000209640, 0x1748da0, 0xc0000b7ec0, 0xc0003058f0, 0xc0001ff3e0, 0x0, 0x0, 0xc0003ffce0, 0x1)
        /home/runner/work/src-cli/src-cli/internal/campaigns/service.go:373 +0x45b
main.campaignsExecute(0x1748da0, 0xc0000b7ec0, 0xc0000b7e80, 0xc000209640, 0xc00020e180, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/runner/work/src-cli/src-cli/cmd/src/campaigns_common.go:218 +0xdd0
main.init.4.func1(0xc0000b8040, 0x8, 0x8, 0x0, 0x0)
        /home/runner/work/src-cli/src-cli/cmd/src/campaigns_preview.go:49 +0x3f7
main.commander.run(0xc000212fc0, 0x5, 0x8, 0xc000200240, 0x16844bc, 0xd, 0x16a85a5, 0x251, 0xc0000b8030, 0x9, ...)
        /home/runner/work/src-cli/src-cli/cmd/src/cmd.go:94 +0x3aa
main.init.1.func1(0xc0000b8030, 0x9, 0x9, 0x9, 0x0)
        /home/runner/work/src-cli/src-cli/cmd/src/campaigns.go:32 +0x97
main.commander.run(0xc00020e200, 0x10, 0x10, 0xc0000b4120, 0x167c7f2, 0x3, 0x16ab01a, 0x46e, 0xc0000b8010, 0xb, ...)
        /home/runner/work/src-cli/src-cli/cmd/src/cmd.go:94 +0x3aa
main.main()
        /home/runner/work/src-cli/src-cli/cmd/src/main.go:68 +0xf7
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,8 @@ All notable changes to `src-cli` are documented in this file.
 
 ### Fixed
 
+- Fix a regression that was introduced by [#361](https://github.com/sourcegraph/src-cli/pull/361) and caused the "Resolving repositories" step of `src campaign [apply|preview]` to crash when the search query in the campaign spec yielded file matches and repository matches from the same repository.
+
 ### Removed
 
 ## 3.22.0

--- a/internal/campaigns/service.go
+++ b/internal/campaigns/service.go
@@ -535,7 +535,11 @@ func (sr *searchResult) UnmarshalJSON(data []byte) error {
 		return nil
 
 	case "Repository":
-		return json.Unmarshal(data, &sr.Repository)
+		if err := json.Unmarshal(data, &sr.Repository); err != nil {
+			return err
+		}
+		sr.Repository.FileMatches = map[string]bool{}
+		return nil
 
 	default:
 		return errors.Errorf("unknown GraphQL type %q", tn.Typename)

--- a/internal/campaigns/service_test.go
+++ b/internal/campaigns/service_test.go
@@ -1,6 +1,14 @@
 package campaigns
 
-import "testing"
+import (
+	"bytes"
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/sourcegraph/src-cli/internal/api"
+)
 
 func TestSetDefaultQueryCount(t *testing.T) {
 	for in, want := range map[string]string{
@@ -20,3 +28,94 @@ func TestSetDefaultQueryCount(t *testing.T) {
 		})
 	}
 }
+
+func TestResolveRepositorySearch(t *testing.T) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/.api/graphql", func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		w.Write([]byte(testResolveRepositorySearchResult))
+	})
+
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+
+	var clientBuffer bytes.Buffer
+	client := api.NewClient(api.ClientOpts{Endpoint: ts.URL, Out: &clientBuffer})
+
+	svc := &Service{client: client}
+
+	repos, err := svc.resolveRepositorySearch(context.Background(), "sourcegraph reconciler or src-cli")
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	// We want multiple FileMatches in the same repository to result in one repository
+	// and Repository and FileMatches should be folded into the same repository.
+	if have, want := len(repos), 3; have != want {
+		t.Fatalf("wrong number of repos. want=%d, have=%d", want, have)
+	}
+}
+
+const testResolveRepositorySearchResult = `{
+  "data": {
+    "search": {
+      "results": {
+        "results": [
+          {
+            "__typename": "FileMatch",
+            "file": { "path": "website/src/components/PricingTable.tsx" },
+            "repository": {
+              "id": "UmVwb3NpdG9yeTo0MDc=",
+              "name": "github.com/sd9/about",
+              "url": "/github.com/sd9/about",
+              "externalRepository": { "serviceType": "github" },
+              "defaultBranch": { "name": "refs/heads/master", "target": { "oid": "1576f235209fbbfc918129db35f3d108347b74cb" } }
+            }
+          },
+          {
+            "__typename": "Repository",
+            "id": "UmVwb3NpdG9yeToxOTM=",
+            "name": "github.com/sd9/src-cli",
+            "url": "/github.com/sd9/src-cli",
+            "externalRepository": { "serviceType": "github" },
+            "defaultBranch": { "name": "refs/heads/master", "target": { "oid": "21dd58b08d64620942401b5543f5b0d33498bacb" } }
+          },
+          {
+            "__typename": "FileMatch",
+            "file": { "path": "cmd/src/api.go" },
+            "repository": {
+              "id": "UmVwb3NpdG9yeToxOTM=",
+              "name": "github.com/sd9/src-cli",
+              "url": "/github.com/sd9/src-cli",
+              "externalRepository": { "serviceType": "github" },
+              "defaultBranch": { "name": "refs/heads/master", "target": { "oid": "21dd58b08d64620942401b5543f5b0d33498bacb" } }
+            }
+          },
+          {
+            "__typename": "FileMatch",
+            "file": { "path": "client/web/src/components/externalServices/externalServices.tsx" },
+            "repository": {
+              "id": "UmVwb3NpdG9yeTo2",
+              "name": "github.com/sourcegraph/sourcegraph",
+              "url": "/github.com/sourcegraph/sourcegraph",
+              "externalRepository": { "serviceType": "github" },
+              "defaultBranch": { "name": "refs/heads/main", "target": { "oid": "e612c34f2e27005928ff3dfdd8e5ead5a0cb1526" } }
+            }
+          },
+          {
+            "__typename": "FileMatch",
+            "file": { "path": "cmd/frontend/internal/httpapi/src_cli.go" },
+            "repository": {
+              "id": "UmVwb3NpdG9yeTo2",
+              "name": "github.com/sourcegraph/sourcegraph",
+              "url": "/github.com/sourcegraph/sourcegraph",
+              "externalRepository": { "serviceType": "github" },
+              "defaultBranch": { "name": "refs/heads/main", "target": { "oid": "e612c34f2e27005928ff3dfdd8e5ead5a0cb1526" } }
+            }
+          }
+        ]
+      }
+    }
+  }
+}
+`

